minimize_size: New `repack` option for qemu-img-based repacking

liori opened the pull request:
I haven't actually tested this on vdi, but based on `qemu-img`'s documentation, it should work just as well.

andsens commented:
Cool! Would it ever make sense to both use vmware-diskmanager *and* qemu-img? Otherwise I'd suggest making `shrink` an enum with the two options and `True/False` as backwards compat.

liori commented:
I don't know, I don't have VMWare installed anywhere. I guess there should be little benefit in running them both. I'll make it into an enum then. Do you prefer squashed merges or should I add the changes as a separate commit?

andsens commented:
Squashed would be best :-)

andsens commented:
Any update on this?

liori commented:
Sorry for the delay. The change is now implemented. Thinking about it, I'd make a small refactoring here and split `shrink.AddRequiredCommands` into three separate tasks for `zerofree` and both types of `shrink` — though in a separate commit or push request, later. Is that desirable?

andsens commented on bootstrapvz/plugins/minimize_size/README.rst:
```diff
@@ -35,8 +39,13 @@ Settings
+   best in conjunction with the zerofree tool. Valid values:
+
+   -  false: Do not shrink.
+   -  ``vmware-vdiskmanage`` or true: Shrink using the `vmware-vdiskmanager`
```
Woops, typo. Missing `r` in `vmware-vdiskmanager`

andsens commented on bootstrapvz/plugins/minimize_size/tasks/shrink.py:
```diff
@@ -8,6 +8,20 @@
 import os
 
 
+def get_shrink_type(plugins):
```
The tasks module should really only contain actual `Task`s. I recommend putting this in `__init__`

andsens commented:
A split-up of the required commands task would be appreciated. Then you don't need the helper function either, you can just inline it in `resolve_tasks()`.

liori commented:
I fixed the typo and added a commit that implements the split of `shrink.AddRequiredCommands`. The helper function is still also used in the `validate_manifest` function, so I didn't inline it; but I moved it to the `minimize_size/__init__.py` file. I'm not sure if the identifiers/description strings for new tasks are proper?

andsens commented:
Looks great! Merging.

The past tense of "shrink" is "shrunk" btw., not that it matters, it's a temp file after all :-)
